File: package/server/middleware.py
#!/usr/bin/env python3

from txg_cli_manager import CommandResult

def to_response(command_result: CommandResult) -> dict:
    """
    Convert raw command execution result tuple into formatted response dict.
    
    Args:
        command_result: Tuple of (stdout, stderr, returncode) from txg_cli.run_command()
        
    Returns:
        Dictionary with processed command results ready for client response
    """
    content, error, returncode, full_command = command_result
    
    # # Clean up output strings
    # cleaned_stdout = stdout.strip() if stdout else ""
    # cleaned_stderr = stderr.strip() if stderr else ""
    
    # TODO: Add filtering logic here later if needed
    # TODO: Add error code normalization if needed
    # TODO: Add logging/metrics if needed
    
    return {
        "content": content,
        "error": error,
        "returncode": returncode,
        "success": returncode == 0,
        "dbg-full-command": full_command  # TODO: Remove this debug field later
    }

def to_analysis_response(command_result: CommandResult) -> dict:
    """
    Convert analysis command result to formatted response with additional guidance.

    Adds helpful message and next steps for analysis jobs that run asynchronously.

    Args:
        command_result: Tuple of (stdout, stderr, returncode) from txg_cli.run_command()

    Returns:
        Dictionary with processed command results plus analysis-specific fields
    """
    response = to_response(command_result)

    if response["success"]:
        response["message"] = "Analysis started successfully. Expected completion time can be several hours depending on data size. The job will continue running even if this conversation ends."
        response["next_steps"] = "Use 'get_analysis_details' tool to check progress. You will receive an email notification when the analysis is complete."

    return response

File: src/middleware.ts
import type { CommandResult } from "./txg-cli-manager.js";

export type ResponseStatus = "success" | "error" | "in_progress";

export interface Response {
  content: string;
  error: string;
  status: ResponseStatus;
  returncode?: number;
  fullCommand: string;
  message?: string;
  next_steps?: string;
}

// Wrap response in MCP protocol structure
// This is separate from toResponse so that:
// 1. Tests can test toResponse logic without MCP wrapper
// 2. If MCP format changes or we switch to MCPServer, only this function changes
export function wrapResponse(response: Response) {
  return {
    content: [
      {
        type: "text",
        text: JSON.stringify(response),
      },
    ],
  };
}

/**
 * Redacts the access token from the command string for security.
 */
function redactAccessToken(command: string): string {
  return command.replace(/--access-token\s+\S+/, "--access-token ***");
}

export function toResponse(result: CommandResult): Response {
  // Handle in-progress result
  if (result.inProgress) {
    return {
      content: (result.stdout || result.stderr).trim(),
      error: "",
      status: "in_progress",
      fullCommand: redactAccessToken(result.fullCommand),
      message:
        "Operation in progress. Depending on file size, this may take a while. The operation will continue in the background.",
    };
  }

  // Handle completed result
  // When command succeeds, treat stderr as additional output, not an error
  // Some CLI tools output success messages to stderr
  const isSuccess = result.exitCode === 0;

  return {
    content: isSuccess
      ? (result.stdout || result.stderr).trim()
      : result.stdout,
    error: isSuccess ? "" : result.stderr,
    status: isSuccess ? "success" : "error",
    returncode: result.exitCode,
    fullCommand: redactAccessToken(result.fullCommand),
  };
}

export function toAnalysisResponse(result: CommandResult): Response {
  const response = toResponse(result);

  if (response.status === "success") {
    return {
      ...response,
      message:
        "Analysis started successfully. Expected completion time can be several hours depending on data size. The job will continue running even if this conversation ends.",
      next_steps:
        "Use 'get_analysis_details' tool to check progress. You will receive an email notification when the analysis is complete.",
    };
  }

  // Return error response as-is
  return response;
}
